### PR TITLE
PMON Voltage and Current Sensor Monitor 

### PR DESCRIPTION
**Pull request by bmridul**
Implementation of SensorMonitor daemon which would provide monitoring of voltage and current sensors.
#### Description

Sensormond daemon is introduced. It collects the voltage and current sensor information from the platform and populates in the StateDB. The sensor data is available to view using CLIs. Sensormond raises syslogs when the sensors report measurement outside the thresholds.

#### Motivation and Context

The change is required for the implementation for the sensor monitoring daemon described in the HLD for voltage and current sensors monitoring.

https://github.com/sonic-net/SONiC/pull/1394

#### How Has This Been Tested?

This is tested with unit test (included in PR) and on target with CLIs and instrumentation of faults.

### CODE REVIEW

File: sonic-sensormond/scripts/sensormond [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+        Update all voltage information to database
+        :return:
+        '''
+        self.log_debug("Start voltage update")
+        for index, voltage_sensor in enumerate(self.chassis.get_all_voltage_sensors()):
+
+            self._refresh_voltage_status(CHASSIS_INFO_KEY, voltage_sensor, index)
+
+        if self.is_chassis_system:
+            available_voltage_sensors = set()
```
Comment (Junchao-Mellanox) [inline sonic-sensormond/scripts/sensormond:185]: Why only handle this for chassis system? For NON chassis system, PSU also can be hot swapable.

File: sonic-sensormond/scripts/sensormond [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+            for tk in table_keys:
+                self.table._del(tk)
+                if self.is_chassis_system and self.chassis_table is not None:
+                    self.chassis_table._del(tk)
+        if self.phy_entity_table:
+            phy_entity_keys = self.phy_entity_table.getKeys()
+            for pek in phy_entity_keys:
+                self.phy_entity_table._del(pek)
+
+    def _log_on_status_changed(self, normal_status, normal_log, abnormal_log):
```
Comment (Junchao-Mellanox) [inline sonic-sensormond/scripts/sensormond:326]: duplicate code

File: sonic-sensormond/scripts/sensormond [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+            if my_slot != INVALID_SLOT:
+                try:
+                    # Modular chassis may not have table CHASSIS_STATE_DB.
+                    table_name = CurrentUpdater.CURRENT_INFO_TABLE_NAME+'_'+str(my_slot)
+                    chassis_state_db = daemon_base.db_connect("CHASSIS_STATE_DB")
+                    self.chassis_table = swsscommon.Table(chassis_state_db, table_name)
+                except Exception as e:
+                    self.chassis_table = None
+
+    def __del__(self):
```
Comment (Junchao-Mellanox) [inline sonic-sensormond/scripts/sensormond:314]: duplicate code

File: sonic-sensormond/scripts/sensormond [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+
+        self.chassis = chassis
+        self.current_status_dict = {}
+        state_db = daemon_base.db_connect("STATE_DB")
+        self.table = swsscommon.Table(state_db, self.CURRENT_INFO_TABLE_NAME)
+        self.phy_entity_table = swsscommon.Table(state_db, PHYSICAL_ENTITY_INFO_TABLE)
+        self.chassis_table = None
+
+        self.is_chassis_system = chassis.is_modular_chassis()
+        if self.is_chassis_system:
```
Comment (Junchao-Mellanox) [inline sonic-sensormond/scripts/sensormond:302]: duplicate code

**Review by Junchao-Mellanox — CHANGES_REQUESTED** [review]

File: sonic-sensormond/scripts/sensormond [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+
+            voltage_status = self.voltage_status_dict[name]
+
+            high_threshold = NOT_AVAILABLE
+            low_threshold = NOT_AVAILABLE
+            high_critical_threshold = NOT_AVAILABLE
+            low_critical_threshold = NOT_AVAILABLE
+            maximum_voltage = NOT_AVAILABLE
+            minimum_voltage = NOT_AVAILABLE
+            voltage = try_get(voltage_sensor.get_value)
```
Comment (Junchao-Mellanox) [inline sonic-sensormond/scripts/sensormond:224]: where is get_unit used?

**Review by Junchao-Mellanox — CHANGES_REQUESTED** [review]

**Comment by bmridul** [comment]
@Junchao-Mellanox , Pls check this and other PRs related to this with the updates.

**Comment by abdosi** [comment]
@Junchao-Mellanox can you please review again
We need to close on this soon.

@bmridul please rebase it as it is out of date

**Review by Junchao-Mellanox — APPROVED** [review]

**Review by abdosi — APPROVED** [review]

**Comment by abdosi** [comment]
@gechiang we would need in msft repo. Since this is enhancement we would need be able to take in 202205 branch.

**Comment by gechiang** [comment]
@abdosi, same issues where there is no msft repo for this submodule and there is no plan for build team to support this new msft repo.  So unless it is a bug fix, it will likely not be able to be taken in for the 202205 at all...

**Comment by Junchao-Mellanox** [comment]
Hi @abdosi , the HLD is not approved, it is in a "Requested Change" status. Should we hold on the merge/cherry-pick until  major comments are addressed?

**Comment by StormLiangMS** [comment]
new feature, no cherry pick

**Comment by shivuv** [comment]
@gechiang @abdosi : Can you please help to cherry pick this into 202305 and 202311 branches? If this is not possible, please advise @bmridul to open new PRs against 202305 and 202311 branches.

**Comment by shivuv** [comment]
@gechiang @abdosi : Can you please help to cherry pick this into 202305 branch?

**Pull request merged**